### ISSUE TITLE
npm postinstall script barfing at compile stage for no apparent reason

### ISSUE THREAD

**Issue by ORESoftware** [npm/npm#14722]
#### I'm opening this issue because:

  - [ x] npm is crashing.
  - [ x] npm is producing an incorrect install.
  - [ x] npm is doing something I don't understand.

#### What's going wrong?

My NPM postinstall node.js script is completely valid TMK, but when NPM runs it we get a Node.js module  compile error

Here is the error I get:

```
> suman@1.1.512188 postinstall /Users/Olegzandr/suman-test/suman-test-projects/node_modules/suman
> node scripts/create-suman-dir.js && node scripts/create-suman-alias.js

/Users/Olegzandr/suman-test/suman-test-projects/node_modules/suman/scripts/create-suman-dir.js:26
    function (cb) {
             ^
SyntaxError: Unexpected token (
    at Object.exports.runInThisContext (vm.js:76:16)
    at Module._compile (module.js:545:28)
    at Object.Module._extensions..js (module.js:582:10)
    at Module.load (module.js:490:32)
    at tryModuleLoad (module.js:449:12)
    at Function.Module._load (module.js:441:3)
    at Module.runMain (module.js:607:10)
    at run (bootstrap_node.js:382:7)
    at startup (bootstrap_node.js:137:9)
    at bootstrap_node.js:497:3
npm ERR! Darwin 15.5.0
npm ERR! argv "/Users/Olegzandr/.nvm/versions/node/v7.0.0/bin/node" "/Users/Olegzandr/.nvm/versions/node/v7.0.0/bin/npm" "install" "-D" "oresoftware/suman"
npm ERR! node v7.0.0
npm ERR! npm  v4.0.2
npm ERR! code ELIFECYCLE

npm ERR! suman@1.1.512188 postinstall: `node scripts/create-suman-dir.js && node scripts/create-suman-alias.js`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the suman@1.1.512188 postinstall script 'node scripts/create-suman-dir.js && node scripts/create-suman-alias.js'.
npm ERR! Make sure you have the latest version of node.js and npm installed.
npm ERR! If you do, this is most likely a problem with the suman package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node scripts/create-suman-dir.js && node scripts/create-suman-alias.js
npm ERR! You can get information on how to open an issue for this project with:
npm ERR!     npm bugs suman
npm ERR! Or if that isn't available, you can get their info via:
npm ERR!     npm owner ls suman
npm ERR! There is likely additional logging output above.

npm ERR! Please include the following file with any support request:
npm ERR!     /Users/Olegzandr/suman-test/suman-test-projects/npm-debug.log
Alexanders-MacBook-Pro:installation Olegzandr$ 

```

this is a completely bizarre problem - NPM postinstall is trying to run my script, but the file is not "compiling" with the core vm module. This JS file runs fine on my machine, in fact here it is, it should give you no compile warnings at the very least:

```js
'use striiiict';

console.log(' => In Suman postinstall script => ', __filename);

//core
const path = require('path');
const fs = require('fs');

//npm
const async = require('async');

//project
const sumanUtils = require('suman-utils/utils');

//cwd
const cwd = process.cwd();

///////////////////////////////////////////////////////////////////////////

const userHomeDir = path.resolve(sumanUtils.getHomeDir());
const p = path.resolve(userHomeDir + '/.suman');
const findSumanExec = path.resolve(p + '/find-local-suman-executable.js');
const sumanDebugLog = path.resolve(p + '/suman-debug.log');
const fileToWrite = fs.readFileSync(require.resolve('../cli/find-local-suman-executable.js'));

fs.mkdir(p, function (err) {

  if (err) {
    if (!String(err.stack || err).match(/EEXIST: file already exists/)) {
      throw err;
    }
  }

  async.parallel([

    function (cb) {
      //always want to update this file to the latest version, so always overwrite
      fs.writeFile(findSumanExec, fileToWrite, { flag: 'w' }, cb);
    },
    function (cb) {
      fs.writeFile(sumanDebugLog, '\n\n => Suman post-install script run on ' + new Date()
        +', from directory (cwd) => ' + cwd, { flag: 'a' }, cb);
    }

  ], function (err) {

    if (err) {
      fs.writeFileSync(sumanDebugLog, '\n => Suman post-install script failed with error => \n' + (err.stack || err), { flag: 'a' });
      throw err;
    }
    else {
      fs.writeFileSync(sumanDebugLog, '\n => Suman post-install script succeeded', { flag: 'a' });
      process.exit(0);
    }

  })

});
```

**Comment by ORESoftware** [comment]
If you run that JS file, like so: "$ node x.js", it should error out, but only _after_ the compile stage.

Any idea what could be happening?

**Comment by ORESoftware** [comment]
Sorry accidentally hit the close button.

**Comment by ORESoftware** [comment]
Doing more research, this bug I mention seems to be a bug with Node version 7.0.0 => when I upgraded to 7.2.0, the problem seems to have gone away. Is this a known bug?

**Comment by legodude17** [comment]
I am not sure, you should go check out the issue tracker of nodejs/node.

**Comment by ORESoftware** [comment]
Yeah my only guess ATM is that the Node.js engine to run the postinstall script was buggy.

**Comment by ORESoftware** [comment]
Yeah, I got this issue again, it is so weird, this should be an NPM issue IMO

**Comment by ORESoftware** [comment]
```
Alexanders-MacBook-Pro:fly-user Olegzandr$ node -c /Users/Olegzandr/WebstormProjects/oresoftware/suman/scripts/create-suman-dir.js
Alexanders-MacBook-Pro:fly-user Olegzandr$ nvm use 4
Now using node v4.3.2 (npm v2.14.12)
Alexanders-MacBook-Pro:fly-user Olegzandr$ node -c /Users/Olegzandr/WebstormProjects/oresoftware/suman/scripts/create-suman-dir.js
Alexanders-MacBook-Pro:fly-user Olegzandr$ nvm use 5
N/A: version "v5" is not yet installed
Alexanders-MacBook-Pro:fly-user Olegzandr$ nvm install 5
Downloading https://nodejs.org/dist/v5.12.0/node-v5.12.0-darwin-x64.tar.gz...
######################################################################## 100.0%
Now using node v5.12.0 (npm v3.8.6)
Alexanders-MacBook-Pro:fly-user Olegzandr$ node -c /Users/Olegzandr/WebstormProjects/oresoftware/suman/scripts/create-suman-dir.js
Alexanders-MacBook-Pro:fly-user Olegzandr$ 
```

**Comment by ORESoftware** [comment]
Just to further prove that the script syntax is valid:

![screenshot 2016-11-24 16 28 29](https://cloud.githubusercontent.com/assets/11139560/20611637/4fa05a8a-b263-11e6-8ee7-31ed66d00d9b.png)

**Comment by legodude17** [comment]
Does it work if you run the same script from terminal?

**Comment by ORESoftware** [comment]
yeah, it does.
What I ended up doing just to try things was to add a hashbang at the top like so

`#!/usr/bin/env node`

and just run it like so

`postinstall: './script.js'`

that works now, but no idea why it was giving me that compile error before, so bizarre.

**Comment by legodude17** [comment]
That is really bizarre, but I am glad you got it working again. What do you
think @KenanY?

**Comment by KenanY** [comment]
Too bizarre for me :laughing:

**Comment by legodude17** [comment]
Should this be closed @KenanY? It is fixed, but we don;t no what the problem is.

**Comment by ORESoftware** [comment]
Yeah for now pls close, if I see the same issue again I will reopen, thank you. I do believe it was a real error, but like I said, I somehow magically solved it as I described above, thanks :)

**Comment by ORESoftware** [comment]
Ughh, I figured out what this was, luckily. The problem went away and then reappeared. It is a git merge issue. I have an automatic publish process and git merge succeeded but mangled the code.

I decided to put in an automatic test as part of the build, to fix this kind of problem, basically

`node -c $(find **/**/*.js)`

then maybe I can catch these potential problems.

Git doesn't play that well with JS, in my experience, there are problems like this one.

node -c is the answer, AFAICT
